File: backspace-compare.py
class Solution(object):
  def backspaceCompare(self, S, T):
    """
    :type S: str
    :type T: str
    :rtype: bool
    """
    p1, p2 = len(S) - 1, len(T) - 1

    while p1 >= 0 or p2 >= 0:
      p1, p2 = self.find_next_letter(S, p1), self.find_next_letter(T, p2)

      if S[p1] != T[p2]: return False

    return True

  def find_next_letter(self, string, pointer):
    num_backspaces = 0

    while pointer >= 0:
      char = string[pointer]
      if char == '#':
        num_backspaces += 1
      elif num_backspaces == 0:
        return pointer - 1
      else:
        num_backspaces -= 1
        
      pointer -= 1

      return None

File: backspace-compare.js
/**
 * @param {string} S
 * @param {string} T
 * @return {boolean}
 */
var backspaceCompare = function (S, T) {
  let p1 = S.length - 1;
  let p2 = T.length - 1;
  let nextS, nextT;

  while (p1 >= 0 && p2 >= 0) {
    [nextS, p1] = findNextLetter(S, p1);
    [nextT, p2] = findNextLetter(T, p2);

    if (nextS !== nextT) return false;
  }

  return true;
};

function findNextLetter(string, pointer) {
  let numBackspaces = 0;
  while (pointer >= 0) {
    let char = string[pointer];
    if (char === '#') {
      numBackspaces++;
    } else if (numBackspaces === 0) {
      return [char, pointer - 1]
    } else {
      numBackspaces--;
    }

    pointer--;
  }

  return [null, -1];
}